From 8a282319bc9eb6dd73557319cbd59d279493cdab Mon Sep 17 00:00:00 2001
From: Ahmad Samir <a.samirh78@gmail.com>
Date: Mon, 4 May 2020 19:49:32 +0200
Subject: [KBookmarkMenu] Assign m_actionCollection early to prevent crash

Summary:
The deprecated ctor that took a KActionCollection param called the new
ctor (that doesn't take an actionCollection before) m_actionCollection was
assigned. This caused the menu actions never to get added to the
actionCollection as it was still nullptr. This caused crashes in
applications that still use the deprecated ctor, e.g. this bug in
konsole https://bugs.kde.org/show_bug.cgi?id=420820.

Since we can't assign m_actionCollection in the initializer list because
then the constructor delegation would follow a member initializer
(info courtsey of the compiler), introduce a helper init method to
reduce code duplication between the two ctor's.

Thanks to Friedrich W. H. Kossebau for figuring it out
https://phabricator.kde.org/D28800#663274.

CCBUG: 420820

Test Plan: make && ctest

Reviewers: #frameworks, dfaure, kossebau, nicolasfella

Reviewed By: dfaure

Subscribers: rikmills, kde-frameworks-devel

Tags: #frameworks

Differential Revision: https://phabricator.kde.org/D29427
---
 src/kbookmarkmenu.cpp | 17 ++++++++++++++---
 src/kbookmarkmenu.h   |  2 ++
 2 files changed, 16 insertions(+), 3 deletions(-)

diff --git a/src/kbookmarkmenu.cpp b/src/kbookmarkmenu.cpp
index 56a4359..41341f6 100644
--- a/src/kbookmarkmenu.cpp
+++ b/src/kbookmarkmenu.cpp
@@ -64,10 +64,18 @@ KBookmarkMenu::KBookmarkMenu(KBookmarkManager *mgr,
                              KBookmarkOwner *_owner,
                              QMenu *_parentMenu,
                              KActionCollection *actionCollection)
-    : KBookmarkMenu(mgr, _owner, _parentMenu)
+    : QObject(),
+      m_actionCollection(actionCollection),
+      d(new KBookmarkMenuPrivate()),
+      m_bIsRoot(true),
+      m_pManager(mgr),
+      m_pOwner(_owner),
+      m_parentMenu(_parentMenu),
+      m_parentAddress(QString())   //TODO KBookmarkAdress::root
 {
-    m_actionCollection = actionCollection;
+    init();
 }
+
 #endif
 
 KBookmarkMenu::KBookmarkMenu(KBookmarkManager* manager, KBookmarkOwner* _owner, QMenu* _parentMenu)
@@ -83,8 +91,11 @@ KBookmarkMenu::KBookmarkMenu(KBookmarkManager* manager, KBookmarkOwner* _owner,
     // TODO KDE5 find a QMenu equvalnet for this one
     //m_parentMenu->setKeyboardShortcutsEnabled( true );
 
-    // qCDebug(KBOOKMARKS_LOG) << "KBookmarkMenu::KBookmarkMenu " << this << " address : " << m_parentAddress;
+    init();
+}
 
+void KBookmarkMenu::init()
+{
     connect(m_parentMenu, &QMenu::aboutToShow,
             this, &KBookmarkMenu::slotAboutToShow);
 
diff --git a/src/kbookmarkmenu.h b/src/kbookmarkmenu.h
index a64bb0e..332a137 100644
--- a/src/kbookmarkmenu.h
+++ b/src/kbookmarkmenu.h
@@ -265,6 +265,8 @@ private Q_SLOTS:
     void slotCustomContextMenu(const QPoint &);
 
 private:
+    void init();
+
     KBookmarkMenuPrivate *d;
 
     bool m_bIsRoot;
-- 
cgit v1.1

